Key-uglifier

AleX04Nov opened the pull request:
This PR adds the key-uglifier demo.
All function names and their functionality are the same as in the source project with the wrong crypto implementation.

AleX04Nov committed:
init key-uglifier

init key-uglifier

vixentael commented on key-uglifier/Cargo.toml:
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "key-uglifier"
+version = "0.1.0"
+edition = "2021"
```
2022?

AleX04Nov replied:
That's not the year, but rather the edition of Rust language: https://doc.rust-lang.org/edition-guide/rust-2021/index.html

vixentael commented on key-uglifier/src/main.rs:
```diff
@@ -0,0 +1,19 @@
+    let mut line: String = String::new();
+    loop {
+        line.clear();
+        println!("Input key: \t");
```
"Input key (32 bytes): \t"

vixentael commented on key-uglifier/src/uglifier.rs:
```diff
@@ -0,0 +1,107 @@
+
+fn read_hex_nibble(c: char) -> (bool, String)
+{
+    if (c >='0') && (c <= '9'){
+        return (true, c.to_string());
+    }
+    else if (c >='A') && (c <= 'F')
+    {
+        return (true, c.to_string());
+    }
+    else if (c >='a') && (c <= 'f')
+    {
+        return (true, c.to_string());
+    }
```
aren't there a simpler way how to check if byte is hex?

smth like https://doc.rust-lang.org/std/primitive.u32.html#method.from_str_radix

or 
https://doc.rust-lang.org/std/primitive.char.html#method.is_digit

AleX04Nov replied:
I tried to save the original logic and flow in those places where I could do it.

vixentael replied:
original to what?

AleX04Nov replied:
According to the source project with the wrong crypto implementation.

Skarlett replied:
> According to the source project with the wrong crypto implementation.

Could you post a reference/link to the source? @AleX04Nov 
If the reference is contained in this repository, post a link with the "lines" (eg. `uri#L10-29`) parameter appended, so its easier to reference what you're eluding to here.

vixentael replied:
> Could you post a reference/link to the source?

@Skarlett no, that's a closed repo.

vixentael reviewed (commented):
nice

G1gg1L3s commented on key-uglifier/src/main.rs:
```diff
@@ -0,0 +1,19 @@
+    loop {
+        line.clear();
+        println!("Input key: \t");
+        std::io::stdin().read_line(&mut line).unwrap();
```
`unwrap` is generally a bad practice for error handling. It's used for cases, where error can indicate some bugs, so we allow closing the process with panics.
Instead, you can use returning error from the main. It's a common practice for application code to just return `Result<(), Box<dyn std::errror::Error>>`, or better to use `anyhow` crate with its `anyhow::Result<()>` type. In this way you specify that you don't care about the type of the error, so if any happens, it will be printed and process will exit with a non-zero code.

Example of usage:
```rust

fn main() -> anyhow::Result<()> {
    // ...
    std::io::stdin().read_line(&mut line)?;
    // ...
    Ok(())
}
```

Here is an article that will help you understand the difference and best practices - https://blog.burntsushi.net/rust-error-handling/

G1gg1L3s commented on key-uglifier/src/uglifier.rs:
```diff
@@ -0,0 +1,107 @@
+fn param_set_default_encryption_key(key_length: u8) -> Vec<u8>
+{
+    let default_key = vec![0x62u8; key_length as usize];
+    return default_key
```
In rust almost everything is an expression, so often `return` is not needed.
```suggestion
    vec![0x62; key_length as usize]
```

G1gg1L3s commented on key-uglifier/src/uglifier.rs:
```diff
@@ -0,0 +1,107 @@
+
+fn convert_to_hex(key_input: &String, key_length: u8) -> (bool, String) {
+    let mut key_output = String::new();
+    let mut bad_chars: bool = false;
```
Types can be derived by the compiler in 90% of the cases. 
```suggestion
    let mut bad_chars = false;
```

G1gg1L3s commented on key-uglifier/src/uglifier.rs:
```diff
@@ -0,0 +1,107 @@
+    let mut key_output = String::new();
+    let mut bad_chars: bool = false;
+
+    for c in key_input.chars().enumerate(){
```
I would suggest you to start using formatters. For example, you can try to setup your ide to format code on save, which is a powerful tool. Or use `cargo fmt`.

G1gg1L3s commented on key-uglifier/src/uglifier.rs:
```diff
@@ -0,0 +1,107 @@
+        let num = read_hex_nibble(c.1);
+        key_output.push_str(&num.1);
+        bad_chars |= !num.0;
+    }
```
You can destruct tuples to name the fields explicitly, to increase readability of the code. For example:
```suggestion
    for (i, c) in key_input.chars().enumerate(){
        if i > ((key_length as usize) * 2) - 1 {
            break;
        }
        let (ok, num) = read_hex_nibble(c);
        key_output.push_str(&num);
        bad_chars |= !ok;
    }
```

G1gg1L3s commented on key-uglifier/src/uglifier.rs:
```diff
@@ -0,0 +1,107 @@
+    return (bad_chars, key_output);
+}
+
+pub(crate) fn param_set_encryption_key(key_input: &String) -> (String, String)
```
It almost never makes sense to pass immutable reference to a string `&String` instead of `&str`. Passing `&String` can cause a double indirection. This is because `&str` under the hood is a (fat) pointer to a an array of bytes. But `&String` is a pointer to a struct with another pointer to a bytes. So, when you write `&String`, cpu will read the pointer from the struct, and then access the byte array.

G1gg1L3s commented on key-uglifier/src/uglifier.rs:
```diff
@@ -0,0 +1,107 @@
+        let bad_chars: bool;
+        (bad_chars, output_key) = convert_to_hex(key_input, key_length);
+        if bad_chars {
+            output_str_desc += &*"\n".to_string();
```
To append a string literal to the `String` you can use either [String::push_str](https://doc.rust-lang.org/src/alloc/string.rs.html#849)
```suggestion
            output_str_desc.push_str("\n");
```

Or in this case, since you are pushing one char, you can use [String::push](https://doc.rust-lang.org/std/string/struct.String.html#method.push) 
```suggestion
            output_str_desc.push('\n');
```

It will prevent unnecessary allocations.

G1gg1L3s commented on key-uglifier/src/uglifier.rs:
```diff
@@ -0,0 +1,107 @@
+        (bad_chars, output_key) = convert_to_hex(key_input, key_length);
+        if bad_chars {
+            output_str_desc += &*"\n".to_string();
+            output_str_desc += &*" WARNING ".to_string().color("black").on_color("bright yellow").bold().to_string();
```
The `Colorize` trait is already [implemented](https://docs.rs/colored/latest/colored/trait.Colorize.html#impl-Colorize-for-%26%27a%20str) for a `&str`, so there is no need to explicitly call `to_string`:
```suggestion
            output_str_desc += &*" WARNING ".color("black").on_color("bright yellow").bold().to_string();
```
But, continuing the previous line, there is also a way to write formatted strings directly to a `String`, similarly how you can write to stdin with `println!`. For this, you have to import `std::fmt::Write` trait:
```suggestion
            use std::fmt::Write;
            // ...
            write!(output_str_desc, "{}", " WARNING ".color("black").on_color("bright yellow").bold()).expect("Write to a string cannot fail");
```

G1gg1L3s reviewed (commented):
Cool!
A lot of small fixes needed, but it's ok for a new rust developer, everyone went this way :) 
I would also suggest you to start using `cargo clippy`. It's a great tool which will help you write _rusty_ code. It will tell not only _what_ but also _why_ something should be changed. For example:
```
$ cargo clippy
   Compiling libc v0.2.126
    Checking lazy_static v1.4.0
    Checking hex v0.4.3
    Checking atty v0.2.14
    Checking colored v2.0.0
    Checking key-uglifier v0.1.0 (/home/gdobn/code/cossacks/blogposts-examples/key-uglifier)
warning: unneeded `return` statement
 --> src/uglifier.rs:7:5
  |
7 |     return default_key
  |     ^^^^^^^^^^^^^^^^^^ help: remove `return`: `default_key`
  |
  = note: `#[warn(clippy::needless_return)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_return

warning: unneeded `return` statement
  --> src/uglifier.rs:13:9
   |
13 |         return (true, c.to_string());
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove `return`: `(true, c.to_string())`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_return

warning: unneeded `return` statement
  --> src/uglifier.rs:17:9
   |
17 |         return (true, c.to_string());
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove `return`: `(true, c.to_string())`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_return

warning: unneeded `return` statement
  --> src/uglifier.rs:21:9
   |
21 |         return (true, c.to_string());
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove `return`: `(true, c.to_string())`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_return

warning: unneeded `return` statement
  --> src/uglifier.rs:26:9
   |
26 |         return (false, '0'.to_string().on_color("green").to_string());
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove `return`: `(false, '0'.to_string().on_color("green").to_string())`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_return

error: this `if` has identical blocks
  --> src/uglifier.rs:12:31
   |
12 |       if (c >='0') && (c <= '9'){
   |  _______________________________^
13 | |         return (true, c.to_string());
14 | |     }
   | |_____^
   |
   = note: `#[deny(clippy::if_same_then_else)]` on by default
note: same as this
  --> src/uglifier.rs:16:5
   |
16 | /     {
17 | |         return (true, c.to_string());
18 | |     }
   | |_____^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#if_same_then_else

warning: manual `RangeInclusive::contains` implementation
  --> src/uglifier.rs:12:8
   |
12 |     if (c >='0') && (c <= '9'){
   |        ^^^^^^^^^^^^^^^^^^^^^^^ help: use: `('0'..='9').contains(&c)`
   |
   = note: `#[warn(clippy::manual_range_contains)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_range_contains

warning: manual `RangeInclusive::contains` implementation
  --> src/uglifier.rs:15:13
   |
15 |     else if (c >='A') && (c <= 'F')
   |             ^^^^^^^^^^^^^^^^^^^^^^^ help: use: `('A'..='F').contains(&c)`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_range_contains

warning: manual `RangeInclusive::contains` implementation
  --> src/uglifier.rs:19:13
   |
19 |     else if (c >='a') && (c <= 'f')
   |             ^^^^^^^^^^^^^^^^^^^^^^^ help: use: `('a'..='f').contains(&c)`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_range_contains

warning: unneeded `return` statement
  --> src/uglifier.rs:43:5
   |
43 |     return (bad_chars, key_output);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove `return`: `(bad_chars, key_output)`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_return

warning: writing `&String` instead of `&str` involves a new object where a slice will do
  --> src/uglifier.rs:30:30
   |
30 | fn convert_to_hex(key_input: &String, key_length: u8) -> (bool, String) {
   |                              ^^^^^^^ help: change this to: `&str`
   |
   = note: `#[warn(clippy::ptr_arg)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg

warning: unneeded `return` statement
   --> src/uglifier.rs:106:5
    |
106 |     return (output_key, output_str_desc);
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove `return`: `(output_key, output_str_desc)`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_return

warning: `key-uglifier` (bin "key-uglifier") generated 11 warnings
error: could not compile `key-uglifier` due to previous error; 11 warnings emitted
```

AleX04Nov committed:
review fixes

Cargo:
-- added `anyhow = "1.0"` crate.

Main:
-- added return type and return value from main;
-- slightly changed print value;
-- added more `quit` strings;
-- changed input error handling;
-- removed explicit types for variables.

Uglifier:
-- changed `return` syntax from functions;
-- changed strings formatting;
-- removed explicit types for variables.

G1gg1L3s commented on key-uglifier/src/uglifier.rs:
```diff
@@ -1,70 +1,55 @@
-    else
-    {
+fn read_hex_nibble(c: char) -> (bool, String) {
+    if "0123456789ABCDEFabcdef".contains(c) {
```
For this purpose, there is a [char::is_digit](https://doc.rust-lang.org/std/primitive.char.html#method.is_digit)

AleX04Nov replied:
No, because `is_digit` checks for `0-9, a-z, A-Z`. But instead of this we need: `0-9, a-f, A-F`

G1gg1L3s replied:
Yeah, it supports `0-9, a-z, A-Z` because it supports $n$-numbering system, where $n$ can be up to 36. But you can specify which numbering system you want.:
```rust
if c.is_digit(16) {
```

G1gg1L3s commented on key-uglifier/src/uglifier.rs:
```diff
@@ -1,70 +1,55 @@
-        bad_chars |= !num.0;
+        let (ok, num) = read_hex_nibble(character);
+        key_output.push_str(&num);
+        bad_chars |= !ok;
```
Personally, I would prefer consistent name of this boolean variable. It's hard to read this expression and keep in mind that _this function_ return boolean that indicates an error, but _that_ indicates ok.

By the way, returning `ok` as a separate bool variable may be replaced by the `Result`. For example:
```rust
fn read_hex_nibble(c: char) -> Result<char, String> {
//...
}

fn convert_to_hex() {
    // ...
    match read_hex_nibble(character) {
        Ok(c) => key_output.push(c),
        Err(err) => {
            key_output.push_str(err);
            bad_chars = true;
        }
    }
}
```

G1gg1L3s reviewed (commented):
Cool :)
Also, I would recommend you to work on your git workflow. It's a good practice to keep a git history clean. Under the clean I mean two things:
- Commit separation - every commit should be atomic change. What does it mean it practice is really hard to explain, you can find a couple of articles about if ([one](https://coderwall.com/p/jmqp0a/why-and-how-i-craft-atomic-commits-in-git) for example). But atomic commit are useful for a number of reasons: they help you to think about changes as separate `steps` (it helps in code review, because a reviewer can see not only the result, but the entire history of how it was achieved), atomic commits can be easily reverted, in case you need to rollback the changes, they also can have a detailed explanation of why such changes are done, keeping all it in the history of project.
- Commit messages - I recommend looking at [this](https://cbea.ms/git-commit/) article. It was the one that inspired me :)

AleX04Nov committed:
changed `read_hex_nibble` return type

Now `read_hex_nibble` returns `Result<String, String>`.
In `convert_to_hex` now checking `read_hex_nibble` in `match` pattern.

AleX04Nov committed:
change validation of char in `read_hex_nibble`

Changed validation of char in `read_hex_nibble` function. Now using `c.is_digit(16)` instead of `"0123456789ABCDEFabcdef".contains(c)`

G1gg1L3s reviewed (approved):
Noice!

vixentael reviewed (approved):
if you have agreed on rust syntax and rules — LGTM
